### PR TITLE
Use ghost locs for nodes used to handle default value of optional arguments

### PR DESCRIPTION
**Pull request by panglesd**
When typing an expression, OCaml turns the parsetree representing:
```ocaml
let f ?(a = true) = 1
```
into the typedtree representing:
```ocaml
let f ?a =
  let a = match a with 
    | Some x -> x
    | None -> true in
  1
```

However, the locations of the `x` pattern and the `x` expression are the one of `true` (in the parsetree), without setting `loc_ghost = true`.

I think `loc_ghost` should be set to `true` for those locations: neither the pattern nor the expression actually are `true`. (This would help tools extracting information from the typedtree: I found this issue while working on code navigation support (merlin-like jump to def) in `odoc`-rendered source.)

### CODE REVIEW

File: typing/typecore.ml [hunk 1]
```diff
@@ -2988,12 +2988,13 @@ and type_expect_
...
-          (Pat.construct ~loc:default_loc
+          (Pat.construct ~loc:default_ghost
              (mknoloc (Longident.(Ldot (Lident "*predef*", "Some"))))
-             (Some ([], Pat.var ~loc:default_loc (mknoloc "*sth*"))))
-          (Exp.ident ~loc:default_loc (mknoloc (Longident.Lident "*sth*")));
+             (Some ([], Pat.var ~loc:default_ghost (mknoloc "*sth*"))))
+          (Exp.ident ~loc:default_ghost (mknoloc (Longident.Lident "*sth*")));
 
         Exp.case
           (Pat.construct ~loc:default_loc
```
Comment (gasche) [inline typing/typecore.ml:3179]: Is it intentional to keep `default_loc` here? Is it the right choice?

File: typing/typecore.ml [hunk 2]
```diff
@@ -2988,12 +2988,13 @@ and type_expect_
       assert(is_optional l); (* default allowed only with optional argument *)
       let open Ast_helper in
       let default_loc = default.pexp_loc in
+      let default_ghost = {default.pexp_loc with loc_ghost = true} in
       let scases = [
```
Comment (gasche) [inline typing/typecore.ml:3171]: Leaving as-is is fine, but I think the code would be easier to read with:

```ocaml
let scases =
  let loc = { default.pexp_loc with loc_ghost = true } in
  [
    ...;
    ...;
  ]
```

**Review by gasche — APPROVED** [review]
Looks reasonable to me, approved in principle. See an inline question about a potential issue left, and a stylistic proposal.

File: typing/typecore.ml (hunk 1 above)
Comment (panglesd) [reply typing/typecore.ml:3179]: Yes, it is intentional: this node corresponds to the default expression in the typedtree (`true` in the example), and I think it is reasonable to leave it to the (non-ghosted) location of the default expression (`true` in the example!)

For instance, if we set it a ghost location, I think Merlin would stop helping on e.g. jump to def inside default expressions, as there would be nothing non-ghost at this location:

```ocaml
let f ?(a = let x = 5 in x) = 1
```

**Comment by gasche** [comment]
Would you like a Change entry for this PR, or did you intentionally not write one?

**Comment by panglesd** [comment]
Yes, I guess a change entry is nice.

I added it in the "Bug fixes" section, but I can move it if it belongs to somewhere else.

**Pull request merged**